File: 01xx/011x/0118-PascalsTriangle/Solution.go
/*  Solution: Calculate nCk without factorials / incrementally
    Time Complexity:    O(n^2) / O(T(n))
    Space Complexity:   O(n^2) / O(T(n))

    Note: T(n) is nth triangular number
 */

func generate(numRows int) [][]int {
  triangle := [][]int{}
  for i := 0; i < numRows; i++ {
    row := []int{1}
    for j := 1; j < i + 1; j++ {
      row = append(row, row[len(row) - 1] * (i - j + 1) / j)
    }
    triangle = append(triangle, row)
  }
  return triangle
}

File: 01xx/011x/0118-PascalsTriangle/Solution.py
# Solution: Binomial Theorem without Factorials
# Time Complexity:  O(T(n))
# Space Complexity: O(T(n))
#
# Note: T(n) is nth triangular number

class Solution:
  def generate(self, numRows: int) -> List[List[int]]:
    triangle = []
    for n in range(numRows):
      row = [1]
      for k in range(1, n+1):
        row.append(int(row[len(row) - 1] * (n-k+1) / k ) )
      triangle.append(row)
    return triangle
